realm/realm-library/src/main/cpp/io_realm_internal_OsObject.cpp

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright 2017 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "io_realm_internal_OsObject.h"
+
+#include <realm/row.hpp>
+#include <object_schema.hpp>
+#include <object.hpp>
+
+#include "util.hpp"
+
+#include "jni_util/java_global_weak_ref.hpp"
+#include "jni_util/java_method.hpp"
+
+using namespace realm;
+using namespace realm::jni_util;
+
+// We need to control the life cycle of Object, weak ref of Java OsObject and the NotificationToken.
+// Wrap all three together, so when the Java object gets GCed, all three of them will be invalidated.
+struct ObjectWrapper {
+    JavaGlobalWeakRef m_row_object_weak_ref;
+    NotificationToken m_notification_token;
+    realm::Object m_object;
+
+    ObjectWrapper(realm::Object& object)
+        : m_row_object_weak_ref()
+        , m_notification_token()
+        , m_object(std::move(object))
+    {
+    }
+
+    ObjectWrapper(ObjectWrapper&&) = delete;
+    ObjectWrapper& operator=(ObjectWrapper&&) = delete;
+
+    ObjectWrapper(ObjectWrapper const&) = delete;
+    ObjectWrapper& operator=(ObjectWrapper const&) = delete;
+
+    ~ObjectWrapper()
+    {
+    }
+};
+
+struct ChangeCallback {
+    ChangeCallback(ObjectWrapper* wrapper)
+        : m_wrapper(wrapper)
+    {
+    }
+
+    void parse_fields(JNIEnv* env, CollectionChangeSet const& change_set)
+    {
+        if (m_field_names_array) {
+            return;
+        }
+
+        if (!change_set.deletions.empty()) {
+            m_deleted = true;
+            return;
+        }
+
+        std::vector<jstring> field_names;
+        auto table = m_wrapper->m_object.row().get_table();
+        for (size_t i = 0; i < change_set.columns.size(); ++i) {
+            if (change_set.columns[i].empty()) {
+                continue;
+            }
+            // FIXME: After full integration of the OS schema, parse the column name from
+            // wrapper->m_object.get_object_schema() will be faster.
+            field_names.push_back(to_jstring(env, table->get_column_name(i)));
+        }
+        m_field_names_array = env->NewObjectArray(field_names.size(), java_lang_string, 0);
```

Comment:
No, this one runs in the thread where the listener registered, so it must be a java thread.

It needs to return a `ObjectArray` with `String`s, every individual entry unfortunately is a java `String` object.